src/codegen/transformations.lisp

```diff
@@ -11,204 +11,191 @@
+                  :node (funcall traverse (node-dynamic-extent-node node) args)
+                  :body (funcall traverse (node-dynamic-extent-body node) args))))
+    (cons (cons ':traverse 'node-bind)
+          (lambda (node traverse &rest args)
+            (declare (type function traverse)
+                     (type node-bind node)
+                     (values node-bind &optional))
+            (make-node-bind
+                  :type (node-type node)
+                  :name (node-bind-name node)
+                  :expr (funcall traverse (node-bind-expr node) args)
+                  :body (funcall traverse (node-bind-body node) args)))))
+   t))
 
-  (:method ((node node-dynamic-extent) funs bound-variables)
-    (declare (type util:symbol-list bound-variables))
-    (let* ((new-bound-variables (cons (node-dynamic-extent-name node) bound-variables))
+(defun optional-call (key funs traverse args node)
+   "Look up a function in `funs` corresponding to `key`, and if it exists, call it with arguments `node`, `traverse` (if `(equal ':traverse (car key))`) and (spread) `args`."
+  (declare (type (cons symbol symbol) key)
+           (type list funs)
+           (type function traverse)
+           (type list args)
+           (type node node))
+  (or
+   (alexandria:when-let*
+       ((f (cdr (assoc key funs :test #'equal))))
+     (if (equal ':traverse (car key))
+         (apply f node traverse args)
+         (apply f node args)))
+   node))
 
-           (node
-             (make-node-dynamic-extent
-              :type (node-type node)
-              :name (node-dynamic-extent-name node)
-              :node (traverse (node-dynamic-extent-node node) funs new-bound-variables)
-              :body (traverse (node-dynamic-extent-body node) funs new-bound-variables))))
-      (call-if node :dynamic-extent funs bound-variables)))
-
-  (:method ((node node-bind) funs bound-variables)
-    (declare (type util:symbol-list bound-variables))
-    (call-if node :before-bind funs bound-variables)
-    (let* ((new-bound-variables (cons (node-bind-name node) bound-variables))
-
-           (node
-             (make-node-bind
-              :type (node-type node)
-              :name (node-bind-name node)
-              :expr (traverse (node-bind-expr node) funs new-bound-variables)
-              :body (traverse (node-bind-body node) funs new-bound-variables))))
-      (call-if node :bind funs bound-variables))))
+(defun traverse (initial-node custom-funs &optional (initial-args nil))
+  "Create a recursive \"generic\" traversal with pre-, in-, and post-operations fixed by `funs` (with the default traversals as a fallback), and evaluate this traversal with `initial-node` and `initial-args`."
+  (declare (type node initial-node)
+           (type list custom-funs)
+           (type list initial-args)
+           (values node &optional))
+  (let ((funs (append custom-funs (make-default-traversals))))
+    (labels ((traverse (current-node args)
```

Comment:
The original `defgeneric traverse` is now entirely gone. Both `funs` and the `traverse` actions are now entirely abstracted to a common alist associating `(cons (or :before :traverse :after) 'coalton-impl/codegen/ast:node*)` to function objects. Instead of passing `funs` everywhere to characterize the traversal, now the fixpoint `traverse` function itself gets passed. Instead of abstracting away the particular `bound-variables` logic and gluing the traversal to itself with some wrapping function, now only the glue remains.